core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -598,14 +599,24 @@ private boolean applyFailedShard(RoutingAllocation allocation, ShardRouting fail
                     while (initializingNode.hasNext()) {
                         ShardRouting shardRouting = initializingNode.next();
                         if (shardRouting.isRelocationTargetOf(failedShard)) {
-                            logger.trace("{} is removed due to the failure of the source shard", shardRouting);
-                            initializingNode.remove();
+                            if (failedShard.primary()) {
+                                logger.trace("{} is removed due to the failure of the source shard", shardRouting);
+                                initializingNode.remove();
+                            } else {
+                                logger.trace("{}, relocation source failed, mark as initializing without relocation source", shardRouting);
+                                removedRelocationSourceOnRelocationTarget = true;
+                                initializingNode.removeRelocationSource();
+                            }
+                            break;
                         }
                     }
                 }
             }
-
-            matchedNode.moveToUnassigned(unassignedInfo);
+            if (removedRelocationSourceOnRelocationTarget) {
```

Comment:
hmm I am not sure I understand why this is skipped. The `matchedNode` is the Routing node that holds the shard that is failed. So we have to move it to unassigned anyway no? I must be missing something - this is a confusion method :)